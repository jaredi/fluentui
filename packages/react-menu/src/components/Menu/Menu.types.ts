import * as React from 'react';
import { ObjectShorthandProps, ShorthandProps } from '@fluentui/react-utilities';
import { MenuListProps } from '../MenuList/index';

/**
 * Extends and drills down Menulist props to simplify API
 * {@docCategory Menu }
 */
export interface MenuProps extends MenuListProps {
  /**
   * Explicitly require children
   */

  children: React.ReactNode;
  /**
   * Whether the popup is open
   */
  open?: boolean;

  /**
   * Whether the popup is open by default
   */
  defaultOpen?: boolean;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;
}

/**
 * {@docCategory Menu }
 */
export interface MenuState extends MenuProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;

  /**
   * Whether the popup is open
   */
  open: boolean;

  /**
   * Callback to open/close the popup
   */
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;

  /**
   * Internal react node that just simplifies handling children
   */
  menuList: React.ReactNode;

  /**
   * Internal react node that just simplifies handling children
   */
  menuTrigger: React.ReactNode;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>>;
}
