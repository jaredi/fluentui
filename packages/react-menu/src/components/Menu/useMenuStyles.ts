import { makeStyles, ax } from '@fluentui/react-make-styles';
import { MenuState } from './Menu.types';

/**
 * Styles for the popup slot
 */
const useMenuPopupStyles = makeStyles<MenuState>([
  [
    null,
    theme => ({
      backgroundColor: theme.alias.color.neutral.neutralBackground1,
      minWidth: '128px',
      minHeight: '48px',
      maxWidth: '300px',
      width: 'fit-content',
      boxShadow: `${theme.alias.shadow.shadow16}`,
      paddingTop: '4px',
      paddingBottom: '4px',
    }),
  ],
]);

/**
 * Apply styling to the Menu slots based on the state
 * {@docCategory Menu }
 */
export const useMenuStyles = (state: MenuState): MenuState => {
  state.menuPopup.className = ax(useMenuPopupStyles(state), state.menuPopup.className);
  return state;
};
