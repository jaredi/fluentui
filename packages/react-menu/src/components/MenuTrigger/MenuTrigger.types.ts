import * as React from 'react';

/**
 * {@docCategory MenuTrigger }
 */
export interface MenuTriggerProps {
  /**
   * Explicitly require single child
   */
  children: React.ReactElement;
}

/**
 * {@docCategory MenuTrigger }
 */
export interface MenuTriggerState extends MenuTriggerProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;

  /**
   * Opens the popup
   */
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
}
