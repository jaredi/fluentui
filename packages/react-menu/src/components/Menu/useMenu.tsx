import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs, useControllableValue } from '@fluentui/react-utilities';
import { MenuProps, MenuState } from './Menu.types';
import { MenuTrigger } from '../MenuTrigger/index';

export const menuShorthandProps: (keyof MenuProps)[] = ['menuPopup'];

const mergeProps = makeMergeProps<MenuState>({ deepMerge: menuShorthandProps });

/**
 * Create the state required to render Menu.
 *
 * The returned state can be modified with hooks such as useMenuStyles,
 * before being passed to renderMenu.
 *
 * @param props - props from this instance of Menu
 * @param ref - reference to root HTMLElement of Menu
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Menu }
 */
export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultProps?: MenuProps): MenuState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      menuPopup: { as: 'div' },
    },
    defaultProps,
    resolveShorthandProps(props, menuShorthandProps),
  );

  // TODO Better way to narrow types ?
  const children = React.Children.toArray(state.children) as React.ReactElement[];

  // TODO throw warnings in development safely
  if (children.length !== 2) {
    // eslint-disable-next-line no-console
    console.warn('Menu can only take one MenuTrigger and one MenuList as children');
  }

  children.forEach(child => {
    if (child.type === MenuTrigger) {
      state.menuTrigger = child;
    } else {
      state.menuList = child;
    }
  });

  state.menuPopup.children = (Component, p) => {
    return <Component {...p}> {state.menuList} </Component>;
  };

  const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
  // TODO fix useControllableValue typing
  state.open = open !== undefined ? open : state.open;
  state.setOpen = React.useCallback(
    (...args) => {
      setOpen(...args);
    },
    [setOpen],
  );

  return state;
};
