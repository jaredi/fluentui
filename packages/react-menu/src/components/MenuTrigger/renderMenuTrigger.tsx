import { MenuTriggerState } from './MenuTrigger.types';

/**
 * Render the final JSX of MenuTrigger
 *
 * Only renders children
 * {@docCategory MenuTrigger }
 */
export const renderMenuTrigger = (state: MenuTriggerState) => {
  return state.children;
};
