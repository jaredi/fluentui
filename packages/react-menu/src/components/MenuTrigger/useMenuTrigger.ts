import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { MenuTriggerProps, MenuTriggerState } from './MenuTrigger.types';
import { useMenuContext, MenuContextValue } from '../../menuContext';

export const menuTriggerShorthandProps: (keyof MenuTriggerProps)[] = [];

const mergeProps = makeMergeProps<MenuTriggerState>({ deepMerge: menuTriggerShorthandProps });

/**
 * Create the state required to render MenuTrigger.
 * Clones the only child component and adds necessary event handling behaviours to open a popup menu
 *
 * @param props - props from this instance of MenuTrigger
 * @param ref - reference to root HTMLElement of MenuTrigger
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory MenuTrigger }
 */
export const useMenuTrigger = (
  props: MenuTriggerProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: MenuTriggerProps,
): MenuTriggerState => {
  const setOpen = useMenuContext(context => context.setOpen);

  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
    },
    defaultProps,
    resolveShorthandProps(props, menuTriggerShorthandProps),
  );

  state.setOpen = setOpen;

  const child = React.Children.only(state.children);
  state.children = React.cloneElement(child as React.ReactElement, getTriggerProps(state.setOpen, child.props));

  return state;
};

// TODO this is quick 'n dirty, follow up and improve interactions
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getTriggerProps = (setOpen: MenuContextValue['setOpen'], props: any) => {
  const triggerProps: React.HTMLAttributes<HTMLElement> = {};
  triggerProps.onClick = e => {
    setOpen(s => !s);
    props.onClick(e);
  };

  return triggerProps;
};
