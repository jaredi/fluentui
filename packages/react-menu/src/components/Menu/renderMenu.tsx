import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { MenuState } from './Menu.types';
import { menuShorthandProps } from './useMenu';
import { MenuProvider } from '../../menuContext';

/**
 * Render the final JSX of Menu
 * {@docCategory Menu }
 */
export const renderMenu = (state: MenuState) => {
  const { slots, slotProps } = getSlots(state, menuShorthandProps);
  const { open, setOpen, onCheckedValueChange, checkedValues, defaultCheckedValues } = state;

  return (
    <MenuProvider
      value={{ open, setOpen, onCheckedValueChange, checkedValues, defaultCheckedValues, hasMenuContext: true }}
    >
      {state.menuTrigger}
      {/** TODO use open state to control a real popup */}
      {state.open && <slots.menuPopup {...slotProps.menuPopup} />}
    </MenuProvider>
  );
};
