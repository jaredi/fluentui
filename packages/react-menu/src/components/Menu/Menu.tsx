import * as React from 'react';
import { useMenu } from './useMenu';
import { MenuProps } from './Menu.types';
import { renderMenu } from './renderMenu';
import { useMenuStyles } from './useMenuStyles';

/**
 * Wrapper component that manages state for a popup MenuList and a MenuTrigger
 * {@docCategory Menu }
 */
export const Menu = React.forwardRef<HTMLElement, MenuProps>((props, ref) => {
  const state = useMenu(props, ref);

  useMenuStyles(state);
  return renderMenu(state);
});

Menu.displayName = 'Menu';
