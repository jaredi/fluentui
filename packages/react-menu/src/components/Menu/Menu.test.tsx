import * as React from 'react';
import { Menu } from './Menu';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { MenuTrigger } from '../MenuTrigger/index';
import { MenuList } from '../MenuList/index';
import { MenuItem } from '../MenuItem/index';

describe('Menu', () => {
  isConformant({
    disabledTests: [
      'as-renders-html',
      'as-renders-fc',
      'component-handles-ref',
      'component-has-root-ref',
      'component-handles-classname',
      'as-passes-as-value',
    ],
    Component: Menu,
    displayName: 'Menu',
    requiredProps: {
      children: [
        <MenuTrigger key="trigger">
          <button>MenuTrigger</button>
        </MenuTrigger>,
        <MenuList key="item">
          <MenuItem>Item</MenuItem>
        </MenuList>,
      ],
    },
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Menu in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(
      <Menu>
        <MenuTrigger>
          <button>Menu trigger</button>
        </MenuTrigger>
        <MenuList>
          <MenuItem>Item</MenuItem>
        </MenuList>
      </Menu>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
