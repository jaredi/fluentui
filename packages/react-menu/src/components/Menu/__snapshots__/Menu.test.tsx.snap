// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders a default state 1`] = `
<button
  onClick={[Function]}
>
  Menu trigger
</button>
`;
